File: android/src/main/java/com/polidea/flutter_ble_lib/event/CharacteristicsMonitorStreamHandler.java
package com.polidea.flutter_ble_lib.event;

import androidx.annotation.NonNull;

import com.polidea.flutter_ble_lib.SingleCharacteristicResponse;
import com.polidea.flutter_ble_lib.constant.ChannelName;
import com.polidea.flutter_ble_lib.converter.BleErrorJsonConverter;
import com.polidea.flutter_ble_lib.converter.SingleCharacteristicResponseJsonConverter;
import com.polidea.multiplatformbleadapter.errors.BleError;

import org.json.JSONException;

import java.util.UUID;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;

public class CharacteristicsMonitorStreamHandler implements EventChannel.StreamHandler {

    @NonNull private EventChannel eventChannel;
    @NonNull public String name;
    @NonNull public String deviceId;

    private EventChannel.EventSink eventSink;
    private SingleCharacteristicResponseJsonConverter characteristicResponseJsonConverter
            = new SingleCharacteristicResponseJsonConverter();
    private BleErrorJsonConverter bleErrorJsonConverter = new BleErrorJsonConverter();

    public CharacteristicsMonitorStreamHandler(BinaryMessenger binaryMessenger, String deviceIdentifier) {
        name = ChannelName.MONITOR_CHARACTERISTIC + "/" + UUID.randomUUID().toString().toUpperCase();
        deviceId = deviceIdentifier;
        eventChannel = new EventChannel(binaryMessenger, name);
        eventChannel.setStreamHandler(this);
    }

    @Override
    synchronized public void onListen(Object o, EventChannel.EventSink eventSink) {
        this.eventSink = eventSink;
    }

    @Override
    synchronized public void onCancel(Object o) {
        eventSink = null;
    }

    synchronized public void onCharacteristicsUpdate(SingleCharacteristicResponse characteristic) throws JSONException {
        if (eventSink != null) {
            eventSink.success(characteristicResponseJsonConverter.toJson(characteristic));
        }
    }

    synchronized public void end() {
        if (eventSink != null) {
            eventSink.endOfStream();
        }
    }

    synchronized public void onError(BleError error, String transactionId) {
        if (eventSink != null) {
            eventSink.error(String.valueOf(error.errorCode.code), error.reason, bleErrorJsonConverter.toJson(error, transactionId));
        }
    }
}

File: android/src/main/kotlin/hu/privatdoktor/flutter_ble_lib/event/CharacteristicsMonitorStreamHandler.kt
package hu.privatdoktor.flutter_ble_lib.event

import android.bluetooth.BluetoothGattCharacteristic
import com.welie.blessed.BluetoothPeripheral
import hu.privatdoktor.flutter_ble_lib.BleError
import hu.privatdoktor.flutter_ble_lib.BleErrorCode
import hu.privatdoktor.flutter_ble_lib.ChannelName
import hu.privatdoktor.flutter_ble_lib.Client
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink

import io.flutter.plugin.common.BinaryMessenger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

import org.json.JSONObject
import java.util.*

class CharacteristicsMonitorStreamHandler(
    binaryMessenger: BinaryMessenger,
    val uniqueKey: String,
) : EventChannel.StreamHandler {
    private val eventChannel = EventChannel(binaryMessenger, uniqueKey)
    private var eventSink: EventSink? = null
    private var cleanUpClosure: (suspend () -> Unit)? = null

    companion object {
        fun generateRandomUniqueKeyFor(
            deviceIdentifier: String,
            char: BluetoothGattCharacteristic
        ) : String {
            val characteristicUuid = char.uuid.toString().lowercase()
            val serviceUuid = char.service.toString().lowercase()
            val unique = UUID.randomUUID().toString().lowercase()
            return "${ChannelName.MONITOR_CHARACTERISTIC}/$deviceIdentifier/$serviceUuid/$characteristicUuid/$unique"
        }
    }

    init {
        eventChannel.setStreamHandler(this)
    }

    override fun onListen(o: Any?, eventSink: EventSink) {
        this.eventSink = eventSink
    }

    override fun onCancel(o: Any?) {
        eventSink = null
        val cleanUpClosure = this.cleanUpClosure
        if (cleanUpClosure != null) {
            GlobalScope.launch(Dispatchers.Main.immediate) {
                cleanUpClosure()
            }
        }
        this.cleanUpClosure = null
    }

    fun onCharacteristicsUpdate(
        peripheral: BluetoothPeripheral,
        serviceUuid: String,
        characteristic: BluetoothGattCharacteristic
    ) {
        val payload = Client.singleCharacteristicWithValueResponse(
            peripheral = peripheral,
            serviceUuidStr = serviceUuid,
            characteristic = characteristic
        )
        try {
            val jsonStr = JSONObject(payload).toString()
            eventSink?.success(jsonStr)
        } catch (e: Throwable) {
            eventSink?.error(
                BleErrorCode.UnknownError.toString(),
                e.localizedMessage,
                e.stackTrace
            )
        }
    }

    fun end() {
        eventSink?.endOfStream()
        eventSink = null
    }

    fun onError(error: BleError) {
        eventSink?.error(
            error.errorCode.code.toString(),
            error.reason,
            error.toJsonString()
        )
    }

    fun afterCancelDo(cleanUpClosure: suspend () -> Unit) {
        this.cleanUpClosure = cleanUpClosure
    }


}